monash-human-power/MHP_DAS
# Add GPS sensor to back WM

## Description

This PR:
- Creates the back WM script, mainly copied off the middle WM code
- Switches to using `asyncio` for anywhere where the code has to wait for something to happen. This allows other code to run instead. (i.e. read data from UART while waiting for MQTT messages to come in)
- Implements the GPS sensor class.
  - I added the `datetime` and [`pdop`](https://en.wikipedia.org/wiki/Dilution_of_precision_(navigation)) fields which aren't on the V3 MQTT spec right now - I reckon they'll be good to have though.

## Screenshots

Here's some example generated data. (Location truncated for privacy as this is a public repo, normally shows to 7 sig figs)

```json
{
  "type": "gps",
  "value": {
    "satellites": 4,
    "pdop": 6.86,
    "latitude": -37.8,
    "longitude": 145,
    "altitude": 140.7,
    "speed": 0.7798978,
    "course": 13.69,
    "datetime": "2021-01-12T12:13:33",
  },
}
```

## Steps to Test

If you have an ESP32 and a compatible GPS module, you should be able to just
- Set the UART channel correctly in `main.py` for however your module is hooked up
- Run `main.py`
- Then, ensure reasonable information is published.
